daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeSpec.scala

```diff
@@ -0,0 +1,329 @@
+
+  val someValue = VersionedValue(v6, left(Value.ValueUnit))
+  val anotherValue = VersionedValue(v6, right(Value.ValueTrue))
+
+  "Node.isReplayedBy" should {
+
+    val someKeyWithM = KeyWithMaintainers(someKey, Set(oscar))
+
+    val create = NodeCreate[ContractId, Val](
+      coid = someContractId,
+      coinst = contract(someValue),
+      optLocation = Some(someLocation),
+      signatories = Set(alice),
+      stakeholders = Set(bob),
+      key = Some(someKeyWithM),
+    )
+
+    val fetch = NodeFetch[ContractId, Val](
+      coid = someContractId,
+      templateId = TemplateTyCon,
+      optLocation = Some(someLocation),
+      actingParties = Some(Set(alice)),
+      signatories = Set(alice),
+      stakeholders = Set(bob),
+      key = Some(someKeyWithM),
+    )
+
+    val exercise = NodeExercises[Nothing, ContractId, Val](
+      targetCoid = someContractId,
+      templateId = TemplateTyCon,
+      choiceId = someChoiceName,
+      optLocation = Some(someLocation),
+      consuming = true,
+      actingParties = Set(alice),
+      chosenValue = someValue,
+      stakeholders = Set(bob),
+      signatories = Set(eve),
+      controllersDifferFromActors = false,
+      children = ImmArray.empty,
+      exerciseResult = Some(someValue),
+      key = Some(someKeyWithM),
+    )
+
+    val lookup = NodeLookupByKey[ContractId, Val](
+      templateId = TemplateTyCon,
+      optLocation = Some(someLocation),
+      key = someKeyWithM,
+      result = Some(someContractId)
+    )
+
+    val aliceAndBob = Set(alice, bob)
+
+    "reject invalid replay of a create node" in {
+
+      val rightTestCases =
+        Table(
+          "create node",
+          fetch,
+          exercise,
+          lookup,
+          create.copy(coid = anotherContractId),
+          create.copy(coinst = create.coinst.copy(template = eitherTyCon)),
+          create.copy(coinst = create.coinst.copy(arg = create.coinst.arg.copy(version = v7))),
+          create.copy(coinst = create.coinst.copy(arg = anotherValue)),
+          create.copy(coinst = create.coinst.copy(agreementText = "disagremenent")),
+          create.copy(signatories = aliceAndBob),
+          create.copy(stakeholders = aliceAndBob),
+          create.copy(key = None),
+          create.copy(key = create.key.map(_.copy(key = anotherKey))),
+          create.copy(key = create.key.map(k => k.copy(key = k.key.copy(version = v7)))),
+          create.copy(key = create.key.map(_.copy(maintainers = aliceAndBob)))
+        )
+
+      val leftTestCases =
+        Table(
+          "recorded node",
+          exercise.copy(key = None)
+        )
+
+      forEvery(rightTestCases)(Node.isReplayedBy(create, _) shouldBe false)
+      forEvery(leftTestCases)(Node.isReplayedBy(_, create) shouldBe false)
+
+    }
+
+    "accept valid replay of a create node" in {
+
+      val leftTestCases =
+        Table(
+          "replayed node",
+          create,
```

Comment:
Do you assume that values have a unique shape? What about a `ValueRecord` where some field names have been omitted, i.e., set to `None`?

Reply:
I don't think it's necessary to test this exhaustively here. But IMO it makes sense to include one positive test case for each kind of node to make sure that we don't use `==` on values inside the nodes.

Reply:
As far I can tell there is no difference between `` `Value Equal instance` `` and standard equality (expect type safety and contract ID parametrization). 
In other words, "Yes we assume that values have a unique shape". Type constructors or record field names can be omitted only for values coming from the ledger API.

Reply:
OK. Makes sense. Then there's no point in testing that here.